src/analysis/calculus/times_cont_diff.lean

```diff
@@ -0,0 +1,749 @@
+field, `n` is the number of iterations, `f` is the function and `x` is the point. We also define a
+version `iterated_fderiv_within` relative to a domain, as well as predicates `times_cont_diff k n f`
+and `times_cont_diff_on k n f s` saying that the function is `C^n`, respectively in the whole space
+or on the set `s`.
+
+We prove basic properties of these notions.
+-/
+
+import analysis.calculus.deriv
+
+noncomputable theory
+local attribute [instance, priority 0] classical.decidable_inhabited classical.prop_decidable
+
+universes u v w
+
+section deriv
+
+open set
+
+variables {k : Type*} [nondiscrete_normed_field k]
+{E : Type u} [normed_group E] [normed_space k E]
+{F : Type u} [normed_group F] [normed_space k F]
+{G : Type u} [normed_group G] [normed_space k G]
+{s s₁ u : set E} {f f₁ : E → F} {f' f₁' : E →L[k] F} {f₂ : E → G}
+{f₂' : E →L[k] G} {g : F → G} {x : E} {c : F}
+{L : filter E} {t : set F} {b : E × F → G} {sb : set (E × F)} {p : E × F}
+{n : ℕ}
+
+include k
+
+/-The n-th derivative of a function belongs to the space E →L[k] (E →L[k] (E ... F)...))),
+where there are n iterations of `E →L[k]`. We define this space inductively, call it
+`iterated_continuous_linear_map k n E F`, and denote it by `E [×n]→L[k] F`. We can define
+it inductively either from the left (i.e., saying that the
+(n+1)-th space S_{n+1} is E →L[k] S_n) or from the right (i.e., saying that
+the (n+1)-th space associated to F, denoted by S_{n+1} (F), is equal to S_n (E →L[k] F)).
+For proofs, it turns out to be more convenient to use the latter approach (from the right),
+as it means to prove things at the (n+1)-th step we only need to understand well enough the
+derivative in E →L[k] F (contrary to the approach from the left, where would would need to know
+enough on the n-th derivative to deduce things on the (n+1)-th derivative).
+In other words, one could define the (n+1)-th derivative either as the derivative of the n-th
+derivative, or as the n-th derivative of the derivative. We use the latter definition.
+
+A difficulty is related to universes: the first and second spaces in the sequence, for n=0
+and 1, are F and E →L[k] F. If E has universe u and F has universe v, then the first one lives in
+v and the second one in max v u. Since they should live in the same universe (as all the other
+spaces in the construction), it means that at the 0-th step we should not use F, but ulift it to
+universe max v u. But we should also ulift its vector space structure and its normed space
+structure. This can certainly be done, but I decided it was not worth it for now. Therefore, the
+definition is only made when E and F live in the same universe.
+-/
+def iterated_continuous_linear_map (k : Type w) [nondiscrete_normed_field k] (n : ℕ)
+  (E : Type u) [normed_group E] [normed_space k E] :
+  ∀(F : Type u) [hF1 : normed_group F] [hF : @normed_space k F _ hF1], Type u :=
+nat.rec_on n (λF hF1 hF, F) (λn aux_n F hF1 hF, by { resetI, exact aux_n (E →L[k] F) })
```

Comment:
A small comment:  it is often easier to read definitions produced with the equation compiler. I tested this one, and it works:
```lean
def iterated_continuous_linear_map (k : Type w) [nondiscrete_normed_field k] : 
  Π (n : ℕ) (E : Type u) [nE : normed_group E] [@normed_space k E _ nE]
    (F : Type u) [hF1 : normed_group F] [hF : @normed_space k F _ hF1], Type u 
| 0     E _ _ F _ _ := F
| (n+1) E _ _ F _ _ := by { resetI, exact iterated_continuous_linear_map n E (E →L[k] F) }
```

Reply:
I have changed this definition, and a few others, to use the equation compiler. Not all of them, though: when the target type is not `Type u` or `Prop`, I could not get it to work (see for instance `iterated_continuous_linear_map.normed_space_rec`). I am possibly missing a trick here, or this is a limitation of the equation compiler.

Reply:
I played with it a little, but I did not have any better luck. I could not get around the mysterious error message, "rec_fn_macro only allowed in meta definitions." So it makes sense to do it by hand.